File: sample-conference-videochat/sample-conference-videochat/Components/Users/Classes/QBDataFetcher/QBDataFetcher.m
//
//  QBDataFetcher.m
//  sample-multiconference-videochat
//
//  Copyright (c) 2017 QuickBlox. All rights reserved.
//

#import "QBDataFetcher.h"
#import "QBCore.h"
#import <Quickblox/Quickblox.h>

static const NSUInteger kQBPageLimit = 50;
static const NSUInteger kQBPageSize = 50;

@implementation QBDataFetcher

+ (void)fetchDialogs:(void(^)(NSArray *dialogs))completion {
    
    NSDictionary *extendedRequest = @{
                                      @"type[in]" : @2,
                                      };
    
    __block void(^t_request)(QBResponsePage *responsePage, NSMutableArray *allDialogs);
    void(^request)(QBResponsePage *, NSMutableArray *) = ^(QBResponsePage *responsePage, NSMutableArray *allDialogs) {
        
        [QBRequest dialogsForPage:responsePage
                  extendedRequest:extendedRequest
                     successBlock:^(QBResponse *response, NSArray *dialogs, NSSet *dialogsUsersIDs, QBResponsePage *page)
         {
             [allDialogs addObjectsFromArray:dialogs];
             
             BOOL cancel = NO;
             page.skip += dialogs.count;
             
             if (page.totalEntries <= page.skip) {
                 
                 cancel = YES;
             }
             
             if (!cancel) {
                 
                 t_request(page, allDialogs);
             }
             else {
                 
                 if (completion != nil) {
                     completion([allDialogs copy]);
                 }
                 
                 t_request = nil;
             }
             
         } errorBlock:^(QBResponse *response) {
             
             if (completion != nil) {
                 completion([allDialogs copy]);
             }
             
             t_request = nil;
         }];
    };
    
    t_request = [request copy];
    NSMutableArray *allDialogs = [[NSMutableArray alloc] init];
    request([QBResponsePage responsePageWithLimit:kQBPageLimit], allDialogs);
}

+ (void)fetchUsers:(void(^)(NSArray *users))completion {
    
    __weak __typeof(self)weakSelf = self;
    __block void(^t_request)(QBGeneralResponsePage *page, NSMutableArray *allUsers);
    void(^request)(QBGeneralResponsePage *, NSMutableArray *) = ^(QBGeneralResponsePage *page, NSMutableArray *allUsers) {
        
        __typeof(weakSelf)strongSelf = weakSelf;
        [QBRequest usersWithTags:Core.currentUser.tags
                            page:page
                    successBlock:^(QBResponse *response, QBGeneralResponsePage *page, NSArray<QBUUser *> *users)
         {
             page.currentPage++;
             [allUsers addObjectsFromArray:users];
             
             BOOL cancel = NO;
             if (page.currentPage * page.perPage >= page.totalEntries) {
                 cancel = YES;
             }
             
             if (!cancel) {
                 t_request(page, allUsers);
             }
             else {
                 
                 if (completion != nil) {
                     completion([strongSelf excludeCurrentUserFromUsersArray:[allUsers copy]]);
                 }
                 
                 t_request = nil;
             }
             
         } errorBlock:^(QBResponse *response) {
             
             if (completion != nil) {
                 completion([strongSelf excludeCurrentUserFromUsersArray:[allUsers copy]]);
             }
             
             t_request = nil;
         }];
    };
    
    t_request = [request copy];
    NSMutableArray *allUsers = [[NSMutableArray alloc] init];
    request([QBGeneralResponsePage responsePageWithCurrentPage:1 perPage:kQBPageSize], allUsers);
}

+ (NSArray *)excludeCurrentUserFromUsersArray:(NSArray *)users {
    
    QBUUser *currentUser = Core.currentUser;
    if ([users containsObject:currentUser]) {
        
        NSMutableArray *mutableArray = [users mutableCopy];
        [mutableArray removeObject:currentUser];
        return [mutableArray copy];
    }
    return users;
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Users/Classes/QBDataFetcher/QBDataFetcher.swift
//
//  QBDataFetcher.swift
//  sample-conference-videochat-swift
//
//  Created by Vladimir Nybozhinsky on 11.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit
import Quickblox

struct DataFetcherConstant {
    static let pageLimit: UInt = 50
    static let pageSize: UInt = 50
}

class QBDataFetcher {
    // MARK: Class Methods
    class func fetchDialogs(_ completion: @escaping (_ dialogs: [QBChatDialog]?) -> Void) {
        let extendedRequest = ["type[in]": "2"]
        var t_request: ((_ responsePage: QBResponsePage?, _ allDialogs: [QBChatDialog]?) -> Void)?
        var allDialogsTempArr = [QBChatDialog]()
        let request: ((QBResponsePage?, [QBChatDialog]?) -> Void)? = { responsePage, allDialogs in
            
            guard let responsePage = responsePage, let allDialogs = allDialogs else { return }
            
            QBRequest.dialogs(for: responsePage, extendedRequest: extendedRequest,
                              successBlock: { response, dialogs, dialogsUsersIDs, page in
                                
                                allDialogsTempArr = allDialogs
                                allDialogsTempArr.append(contentsOf: dialogs)
                                page.skip += dialogs.count
                                let isLastPage = page.totalEntries <= page.skip
                                let cancel = isLastPage ? true : false
                                if cancel == false {
                                    t_request?(page, allDialogsTempArr)
                                } else {
                                    completion(allDialogsTempArr)
                                    t_request = nil
                                }
                                
            }, errorBlock: { response in
                completion(allDialogsTempArr)
                t_request = nil
            })
        }
        t_request = request
        let allDialogs: [QBChatDialog] = []
        request?(QBResponsePage(limit: Int(DataFetcherConstant.pageLimit)), allDialogs)
    }
    
    class func fetchUsers(_ completion: @escaping (_ users: [QBUUser]?) -> Void) {
        var t_request: ((_ page: QBGeneralResponsePage?, _ allUsers: [QBUUser]?) -> Void)?
        var allUsersTempArray = [QBUUser]()
        let request: ((QBGeneralResponsePage?, [QBUUser]?) -> Void)? = { page, allUsers in
            
            guard let allUsers = allUsers, let currentUserTags = Core.instance.currentUser?.tags else {
                return }
            
            QBRequest.users(withTags: currentUserTags, page: page,
                            successBlock: { response, page, users in
                                
                                page.currentPage = page.currentPage + 1
                                allUsersTempArray = allUsers
                                allUsersTempArray.append(contentsOf: users)
                                let isLastPage = page.currentPage * page.perPage >= page.totalEntries
                                let cancel = isLastPage ? true : false
                                if cancel == false {
                                    t_request?(page, allUsersTempArray)
                                } else {
                                    completion(self.excludeCurrentUser(fromUsersArray: allUsersTempArray))
                                    t_request = nil
                                }
                                
            }, errorBlock: { response in
                completion(self.excludeCurrentUser(fromUsersArray: allUsersTempArray))
                t_request = nil
            })
        }
        t_request = request
        let allUsers: [QBUUser] = []
        request?(QBGeneralResponsePage(currentPage: 1, perPage: DataFetcherConstant.pageSize), allUsers)
    }
    
    class func excludeCurrentUser(fromUsersArray users: [QBUUser]?) -> [QBUUser]? {
        let currentUser = Core.instance.currentUser
        if let users = users {
            let contains = users.contains(where: {$0 == currentUser})
            if contains {
                let mutableArray = users
                return mutableArray.filter({$0 != currentUser})
            }
        }
        return users
    }
}
